sql: add ColumnGeneratedAsIdentity as a new schema change element

shghasemi opened the pull request:
Previously, GeneratedAsIdentity was part of the Column element.
Creating a new schema change element will allow the declarative schema changer
to add/remove/update such columns without dropping the Column element.

Epic: [CRDB-31283](https://cockroachlabs.atlassian.net/browse/CRDB-31283)

Part of: https://github.com/cockroachdb/cockroach/issues/142918

Release note: None

shghasemi committed:
sql: add ColumnGeneratedAsIdentity as a new schema change element

Previously, GeneratedAsIdentity was part of the Column element.
Creating a new schema change element will allow the declarative schema changer
to add/remove/update such columns without dropping the Column element.

Epic: CRDB-31283

Part of cockroachdb#142918

Release note: None

fqazi reviewed (changes requested):
Great work @shghasemi! A couple of minor issues

> @fqazi reviewed 25 of 47 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-:-OdKA46kB5VRQMBsESDg:biosp5o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shghasemi)

___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_generated_as_identity.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9RVu8XZCwcc_qqTy:-OdK9RVu8XZCwcc_qqTz:b4esvk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_generated_as_identity.go#L25)):*
> ```Go
> 				}),
> 			),
> 			equiv(scpb.Status_WRITE_ONLY),
> ```

Let's have a comment explaining why we need this, which is only for the migration case.
___
*[`pkg/sql/schemachanger/screl/attr.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9mVZE0PHFJw7yx6m:-OdK9mVZE0PHFJw7yx6n:b-bjwyom) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/screl/attr.go#L120)):*
> ```Go
> 	// GeneratedAsIdentityType is the type for a generated as identity column.
> 	// It's value must be in catpb.GeneratedAsIdentityType.
> 	GeneratedAsIdentityType
> ```

The rules are not using these fields. We don't need any new attrs here.
___
*[`pkg/sql/schemachanger/screl/attr.go` line 335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9w2s7zEwB7AsFzW5:-OdK9w2s7zEwB7AsFzW6:bcfod1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/screl/attr.go#L335)):*
> ```Go
> 		rel.EntityAttr(DescID, "TableID"),
> 		rel.EntityAttr(ColumnID, "ColumnID"),
> 		rel.EntityAttr(GeneratedAsIdentityType, "Type"),
> ```

Remove the type and sequence opts unless the rules need them.
___
*[`pkg/sql/schemachanger/scpb/migration_test.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdKA3Mh6f1F9ahyrSFz:-OdKA3Mh6f1F9ahyrSG-:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scpb/migration_test.go#L143)):*
> ```Go
> // TestDeprecatedColumnGeneratedAsIdentity will ensure that a ColumnGeneratedAsIdentity
> // element is added.
> func TestDeprecatedColumnGeneratedAsIdentity(t *testing.T) {
> ```

Nice test!
___
*[`pkg/sql/schemachanger/scplan/internal/rules/release_25_4/helpers.go` line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9cI_DRIuC8D0wGJc:-OdK9cI_DRIuC8D0wGJd:b-2yb2gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scplan/internal/rules/release_25_4/helpers.go#L257)):*
> ```Go
> 		return true
> 	case *scpb.ColumnGeneratedAsIdentity:
> 		return true
> ```

Leave the older rules the same.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/release_25_2/helpers.go` line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9_FwDC8-GoOoGp81:-OdK9_FwDC8-GoOoGp82:b-2yb2gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scplan/internal/rules/release_25_2/helpers.go#L215)):*
> ```Go
> 		return true
> 	case *scpb.ColumnGeneratedAsIdentity:
> 		return true
> ```

Leave the older rules the same.
___
*[`pkg/sql/schemachanger/rel/ordinal_set.go` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK8aUY1OQ5TSXpCNfF:-OdK8aUY1OQ5TSXpCNfG:b-j79zi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/rel/ordinal_set.go#L15)):*
> ```Go
> // ordinal is used to correlate attributes in a schema.
> // It enables use of the ordinalSet.
> type ordinal uint16
> ```

What is this change here?
___
*[`pkg/sql/schemachanger/scplan/internal/rules/release_25_3/helpers.go` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9bg49ZRCHhN6C7ke:-OdK9bg49ZRCHhN6C7kf:b-2yb2gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scplan/internal/rules/release_25_3/helpers.go#L221)):*
> ```Go
> 		return true
> 	case *scpb.ColumnGeneratedAsIdentity:
> 		return true
> ```

Leave the older rules the same.
___
*[`pkg/sql/schemachanger/scpb/migration.go` line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9IL01Qxc5EbCHqmc:-OdK9IL01Qxc5EbCHqmd:b-f6wvfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scpb/migration.go#L82)):*
> ```Go
> 	// Migrate GeneratedAsIdentity from column to a separate element
> 	if column := target.GetColumn(); column != nil {
> 		if version.IsActive(clusterversion.V26_1) &&
> ```

Nit: Lets just && this above to reduce nesting.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
> @fqazi reviewed 22 of 47 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156547)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shghasemi)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/release_25_3/helpers.go` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdK9bg49ZRCHhN6C7ke:-OdKL8qsCUfJFUNjdiIo:bkm7vzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/scplan/internal/rules/release_25_3/helpers.go#L221)):*
<details><summary><i>Previously, shghasemi wrote…</i></summary><blockquote>

The following test fails without this change:
`FAIL: //pkg/sql/schemachanger/scplan/internal/rules/release_25_3:release_25_3_test`

</blockquote></details>

The test is incorrectly using ForEachElementType instead of ForEachElementInActiveVersion
___
*[`pkg/sql/schemachanger/rel/rel_test.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-OdKLWfpBRyGDvtc0PaR:-OdKLWfpBRyGDvtc0PaS:b-677437) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fc3c0c99c0f19254dce6c0faf4f1a3523877cb8/pkg/sql/schemachanger/rel/rel_test.go#L184)):*
> ```Go
> 		F10, F11, F12, F13, F14, F15, F16, F17, F18, F19 *uint32
> 		F20, F21, F22, F23, F24, F25, F26, F27, F28, F29 *uint32
> 		F30, F31, F32, F33, F34, F35, F36, F37, F38, F39 *uint32
> ```

Lets revert this change too


<!-- Sent from Reviewable.io -->

shghasemi committed:
sql: Fix TestRuleAssertions to only check active version elements

Previously, TestRuleAssertions was using ForEachLementType instead of
ForEachElementInActiveVersion. This caused invalid test fails when
a new element was added in a newer version.

Epic: CRDB-31283

Release note: None

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 55 of 55 files at r2, 47 of 47 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156547#-:-OdKcjtDDE1AE7STBP1M:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @shghasemi)



<!-- Sent from Reviewable.io -->
